### PR TITLE
simplify+unify offset.apply logic

### PR DESCRIPTION
**Pull request by jbrockmendel**
Implement _get_lastbday mirroring _get_firstbday.

Implement bstart and bend cases in shift month.  Use these to simplify a bunch of the BusinessFoo offsets.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Running asv now.

**Comment by jbrockmendel** [comment]
```
taskset 6 asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries -b timedelta -b period
[...]
       before           after         ratio
     [7f4c960c]       [f6aebe37]
+           44.3s            1.07m     1.45  gil.nogil_datetime_fields.time_period_to_datetime
+           15.0s            17.1s     1.14  gil.nogil_datetime_fields.time_datetime_to_period
+         760±2ns          851±5ns     1.12  period.Properties.time_daysinmonth
-      30.6±0.3μs       27.8±0.2μs     0.91  timeseries.Offsets.time_custom_bday_decr
-     1.87±0.06ms      1.66±0.01ms     0.89  timeseries.DatetimeIndex.time_add_timedelta
-         811±5ns          720±4ns     0.89  period.Properties.time_day
-         836±8ns          736±3ns     0.88  period.PeriodProperties.time_hour
-         204±2ns        154±0.1ns     0.76  timedelta.TimedeltaProperties.time_timedelta_days
-        191±10ms       27.2±0.2ms     0.14  timeseries.AsOfDataFrame.time_asof_nan
-        202±10ms       24.2±0.3ms     0.12  timeseries.AsOfDataFrame.time_asof

taskset 6 asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries -b timedelta -b period
[...]
       before           after         ratio
     [7f4c960c]       [f6aebe37]
+      15.8±0.6μs       20.2±0.1μs     1.28  timeseries.Offsets.time_custom_bday_apply
+      27.1±0.2μs       32.8±0.1μs     1.21  timeseries.Offsets.time_custom_bday_cal_decr
+         725±5ns          833±9ns     1.15  period.PeriodProperties.time_minute
+     16.7±0.08μs       19.0±0.1μs     1.14  timeseries.AsOf.time_asof_single_early
+      17.9±0.1μs       20.3±0.3μs     1.13  timeseries.Offsets.time_timeseries_day_apply
+         102±2μs          114±1μs     1.12  timeseries.Offsets.time_custom_bmonthend_decr_n
+       134±0.9ms          150±1ms     1.12  timeseries.ToDatetime.time_iso8601_tz_spaceformat
+     18.9±0.06μs       21.0±0.2μs     1.11  timeseries.SemiMonthOffset.time_end_incr_n
-           44.6s            39.6s     0.89  gil.nogil_datetime_fields.time_period_to_datetime
-      20.3±0.1μs       17.0±0.1μs     0.84  timeseries.Offsets.time_custom_bday_apply_dt64
-        28.3±2ms           15.7ms     0.56  timeseries.DatetimeIndex.time_dti_tz_factorize
-        28.1±1ms       6.93±0.6ms     0.25  timeseries.DatetimeIndex.time_dti_factorize

taskset 6 asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries -b timedelta -b period
[...]
       before           after         ratio
     [7f4c960c]       [f6aebe37]
+           14.5s            17.0s     1.17  gil.nogil_datetime_fields.time_datetime_to_period
-      8.14±0.5ms      7.00±0.05ms     0.86  timeseries.DatetimeIndex.time_add_offset_fast
-        29.1±1ms       6.95±0.5ms     0.24  timeseries.DatetimeIndex.time_dti_tz_factorize
-        33.2±5ms       6.44±0.4ms     0.19  timeseries.DatetimeIndex.time_dti_factorize
-         301±7ms         57.2±1ms     0.19  timeseries.AsOfDataFrame.time_asof
-         318±9ms       59.7±0.6ms     0.19  timeseries.AsOfDataFrame.time_asof_nan

taskset 6 asv continuous -f 1.1 -E virtualenv master HEAD -b timeseries -b timedelta -b period
[...]
      before           after         ratio
     [7f4c960c]       [f6aebe37]
+     12.0±0.04μs      13.4±0.08μs     1.12  timeseries.SemiMonthOffset.time_begin_apply
+      10.5±0.1μs      11.6±0.02μs     1.11  timeseries.Offsets.time_timeseries_year_incr
-      28.0±0.1μs      24.1±0.07μs     0.86  timeseries.Offsets.time_custom_bday_cal_incr
-      29.5±0.6ms       6.74±0.3ms     0.23  timeseries.DatetimeIndex.time_dti_tz_factorize

```

The optimist in me thinks the time_dti_tz_factorize might be real, but for the most part this looks like a wash.

**Comment by jbrockmendel** [comment]
```
       before           after         ratio
     [7f4c960c]       [f6aebe37]
+      46.2±0.4ms       51.8±0.6ms     1.12  timedelta.ToTimedelta.time_convert_string
-           2.37s            2.02s     0.85  timeseries.AsOfDataFrame.time_asof_nan
-      18.4±0.1μs      15.6±0.05μs     0.85  timeseries.SemiMonthOffset.time_end_incr
-       170±0.4ms        138±0.6ms     0.81  timeseries.ToDatetime.time_iso8601_tz_spaceformat
```

**Comment by jbrockmendel** [comment]
These numbers are against an out-of-date master.  Will update.

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -139,11 +139,21 @@ def apply_index_wraps(func):
...
+
+cpdef int _get_firstbday(int wkday, int days_in_month=0):
+    """
+    (wkday, days_in_month) is the result of monthrange(year, month)
 
     If it's a saturday or sunday, increment first business day to reflect this
+
+    days_in_month arg is a dummy so that this has the same signature as
+    _get_lastbday.
     """
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: you can type these

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -416,15 +425,21 @@ cpdef datetime shift_month(datetime stamp, int months, object day_opt=None):
         dy -= 1
     year = stamp.year + dy
 
-    dim = monthrange(year, month)[1]
+    (wkday, days_in_month) = monthrange(year, month)
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: no parens

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -416,15 +425,21 @@ cpdef datetime shift_month(datetime stamp, int months, object day_opt=None):
...
     if day_opt is None:
-        day = min(stamp.day, dim)
+        day = min(stamp.day, days_in_month)
     elif day_opt == 'start':
         day = 1
     elif day_opt == 'end':
-        day = dim
+        day = days_in_month
+    elif day_opt == 'bstart':
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: how about business_start/end

File: pandas/tseries/offsets.py [hunk 4]
```diff
@@ -1693,21 +1676,15 @@ class BQuarterEnd(QuarterOffset):
...
         n = self.n
-        base = other
-        other = datetime(other.year, other.month, other.day,
-                         other.hour, other.minute, other.second,
-                         other.microsecond)
 
         wkday, days_in_month = tslib.monthrange(other.year, other.month)
-        lastBDay = days_in_month - max(((wkday + days_in_month - 1)
-                                        % 7) - 4, 0)
+        lastBDay = _get_lastbday(wkday, days_in_month)
```
Comment (jreback) [inline pandas/tseries/offsets.py]: ok to de-privatize ``get_lastbday/first``

**Review by jreback — CHANGES_REQUESTED** [review]
small comments. nice removing code! can you confirm that we have sufficient test coverage of things that you are changing (I think yes, but have a look)

File: pandas/_libs/tslibs/offsets.pyx [hunk 5]
```diff
@@ -139,12 +139,23 @@ def apply_index_wraps(func):
 # ---------------------------------------------------------------------
 # Business Helpers
 
-cpdef int _get_firstbday(int wkday):
+cpdef int get_lastbday(int wkday, int days_in_month):
     """
-    wkday is the result of monthrange(year, month)
+    (wkday, days_in_month) is the result of monthrange(year, month)
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: can you update doc-strings

**Review by jreback — CHANGES_REQUESTED** [review]
also can you unit test some these?

File: pandas/_libs/tslibs/offsets.pyx (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx]: Update to reflect what?

**Comment by jbrockmendel** [comment]
> also can you unit test some these?

Yes, but since I just added a test_liboffsets file in #18280 I'd rather close these out separately and then add unit tests in a shared follow-up.

**Comment by jorisvandenbossche** [comment]
@jbrockmendel general comment: can you try to use a bit more descriptive PR titles? "Tslibs offsets paramd" does not say much to me :-)

**Comment by jbrockmendel** [comment]
> can you try to use a bit more descriptive PR titles

You got it.

File: pandas/_libs/tslibs/offsets.pyx (hunk 5 above)
Comment (jreback) [reply pandas/_libs/tslibs/offsets.pyx]: I mean update to a more proper doc-string

**Review by jreback — CHANGES_REQUESTED** [review]
can you add some unit tests

**Comment by jbrockmendel** [comment]
> I mean update to a more proper doc-string

Sure thing.

**Comment by jbrockmendel** [comment]
pls hold off on this pending double-checking cross-DST behavior.

**Comment by jbrockmendel** [comment]
OK, I've convinced myself the DST issue is handled.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks, as always more tests are welcome! (w.r.t. DST)

**Comment by jbrockmendel** [comment]
> thanks, as always more tests are welcome! (w.r.t. DST)

I'm psyched to have this merged.  There's some follow-ups headed you're way I think you'll like.

w/r/t DST `shift_month` (and `shift_months`) is more fragile than I'd like, are safe largely because all existing usages are wrapped with `apply_wraps`.  But I don't want to rely on that always being the case.  So expect tests+fixes.

Specifically w/r/t to DST tests I've been thinking of trying hypothesis (#17978)

**Pull request merged**